### PR TITLE
Add sBOM functionality

### PR DESCRIPTION
**Pull request by aemengo**
Add sBOM functionality

* Opt for `/layers/sbom` instead of `/layers/config/sbom`, because everything in `./config` is stored on the final app image
* Merging of CycloneDX BOMs functionality not part of this PR, since still waiting on implementation from **CycloneDX/cyclonedx-go** library.
* Opt for `application/spdx+json` over `text/spdx+json`
   in virtue of https://www.iana.org/assignments/media-types/application/spdx+jso

Fixes:
#738
#736
#735
#733
#751
#755

### CODE REVIEW

File: buildpack/bomfile.go [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+	LayerName   string
+	LayerType   LayerType
+	Path        string
+}
+
+func (b *BOMFile) Name() string {
+	name := filepath.Base(b.Path)
+
+	switch {
+	case strings.HasSuffix(name, "bom.cdx.json"):
```
[the remark this answers is not in the document]
Comment (natalieparellano) [reply buildpack/bomfile.go:29]: Is this something that we want to add?

File: buildpack/bomfile.go [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+type LayerType int
+
+type BOMFile struct {
+	BuildpackID string
+	LayerName   string
+	LayerType   LayerType
+	Path        string
+}
+
+func (b *BOMFile) Name() (string, bool) {
```
Comment (natalieparellano) [inline buildpack/bomfile.go:25]:
```suggestion
func (b *BOMFile) BOMType() (string, bool) {
```
This seems more accurate... could we clarify what the second return value is supposed to mean?

File: cmd/lifecycle/restorer.go [hunk 3]
```diff
@@ -4,6 +4,11 @@ import (
 	"errors"
 	"fmt"
 
+	"github.com/buildpacks/imgutil"
+	"github.com/buildpacks/imgutil/local"
+	"github.com/buildpacks/imgutil/remote"
+	"github.com/docker/docker/client"
+
```
Comment (natalieparellano) [inline cmd/lifecycle/restorer.go:11]:
```suggestion
```
(need to sort the resulting block)

File: platform/platform.go [hunk 4]
```diff
@@ -3,6 +3,8 @@ package platform
 import (
 	"fmt"
 
+	v08 "github.com/buildpacks/lifecycle/platform/v08"
+
```
Comment (natalieparellano) [inline platform/platform.go:7]:
```suggestion
```
(need to sort the resulting block)

File: restorer.go [hunk 5]
```diff
@@ -93,16 +101,47 @@ func (r *Restorer) Restore(cache Cache) error {
 			}
 		}
 	}
+
+	if api.MustParse(r.Platform.API()).AtLeast("0.8") {
+		if r.LayersMetadata.BOM != nil && r.LayersMetadata.BOM.SHA != "" {
+			g.Go(func() error {
+				return r.restorePreviousLayer(r.LayersMetadata.BOM.SHA)
```
Comment (natalieparellano) [inline restorer.go:108]: It's kind of weird that `PreviousImage` lives on the restorer whereas `cache` is passed in. Maybe they should both be passed in (because they are both sources of things to restore)?

File: buildpack/bomfile.go (hunk 2 above)
Comment (natalieparellano) [reply buildpack/bomfile.go:25]: Or maybe `BOMFormat`?

Comment (natalieparellano) [reply buildpack/bomfile.go:25]: Actually maybe `Name` is fine. Perhaps we could add a docs comment to make it clearer.

File: cache.go [hunk 6]
```diff
@@ -73,3 +79,27 @@ func (e *Exporter) addOrReuseCacheLayer(cache Cache, layerDir layerDir, previous
 	e.Logger.Debugf("Layer '%s' SHA: %s\n", layer.ID, layer.Digest)
 	return layer.Digest, cache.AddLayerFile(layer.TarPath, layer.Digest)
 }
+
+func (e *Exporter) addSBOMCacheLayer(layersDir string, cacheStore Cache, origMetadata platform.CacheMetadata, meta *platform.CacheMetadata) error {
+	// Look for sBOM cache layer
```
Comment (natalieparellano) [inline cache.go:84]:
```suggestion
```
(redundant)

File: cache.go [hunk 7]
```diff
@@ -73,3 +79,27 @@ func (e *Exporter) addOrReuseCacheLayer(cache Cache, layerDir layerDir, previous
...
+	sbomLayer, err := readLayersConfigSBOM(layersDir, "cache", e.Logger)
```
Comment (natalieparellano) [inline cache.go:85]:
```suggestion
	sbomDir, err := readLayersConfigSBOM(layersDir, "cache", e.Logger)
```
(`Layer` makes me think we're looking for a tarball)

Comment (aemengo) [reply cache.go]: _Fixed in latest commit._

File: cache.go (hunk 6 above)
Comment (aemengo) [reply cache.go]: _Fixed in latest commit._

File: restorer.go (hunk 5 above)
Comment (aemengo) [reply restorer.go]: _Fixed in latest commit._

File: platform/platform.go (hunk 4 above)
Comment (aemengo) [reply platform/platform.go]: _Fixed in latest commit._

File: cmd/lifecycle/restorer.go (hunk 3 above)
Comment (aemengo) [reply cmd/lifecycle/restorer.go]: _Fixed in latest commit._

File: buildpack/bomfile.go (hunk 2 above)
Comment (aemengo) [reply buildpack/bomfile.go:30]: _Fixed in latest commit._

File: builder.go [hunk 8]
```diff
@@ -103,6 +107,12 @@ func (b *Builder) Build() (*platform.BuildMetadata, error) {
 		}
 	}
 
+	b.Logger.Debug("Copying BOM files")
+	err = b.copyBOMFiles(config.LayersDir, bomFiles)
```
Comment (natalieparellano) [inline builder.go:111]: Do we want to put the platform API gate here, to make it more readable?

File: buildpack/build_test.go [hunk 9]
```diff
@@ -229,36 +199,113 @@ func testBuild(t *testing.T, when spec.G, it spec.S) {
...
-						Labels:      []buildpack.Label{},
-						MetRequires: []string{"some-deprecated-bp-replace-version-dep", "some-dep", "some-replace-version-dep"},
-						Processes:   []launch.Process{},
-						Slices:      []layers.Slice{},
-					}); s != "" {
-						t.Fatalf("Unexpected:\n%s\n", s)
+					}, br)
+				})
+
+				it("does not process BOM files for old BP API versions", func() {
```
Comment (natalieparellano) [inline buildpack/build_test.go:286]:
```suggestion
				it("does not include BOM files for old BP API versions", func() {
```
```suggestion
				it("does not process BOM files for old BP API versions", func() {
```

File: platform/v08/exit.go [hunk 10]
```diff
@@ -0,0 +1,35 @@
+package v08
+
+import (
+	"github.com/buildpacks/lifecycle/platform/common"
+)
+
+var exitCodes = map[common.LifecycleExitError]int{
```
Comment (natalieparellano) [inline platform/v08/exit.go:7]: This makes me regret the `platform` interface and trying to shovel all platform-related logic into there. But that's another story...

File: buildpack/build_test.go [hunk 11]
```diff
@@ -229,36 +199,113 @@ func testBuild(t *testing.T, when spec.G, it spec.S) {
...
+							Slices:    []layers.Slice{},
+						}); s != "" {
+							t.Fatalf("Unexpected:\n%s\n", s)
+						}
+						assertLogEntry(t, logHandler, "BOM table isn't supported in this buildpack api version. The BOM should be written to <layer>.bom.<ext>, launch.bom.<ext>, or build.bom.<ext>.")
+					})
+				})
+
+				it("process any created BOM files", func() {
+					buildpackID := "A"
```
Comment (natalieparellano) [inline buildpack/build_test.go:236]:
```suggestion
					buildpackID := bpTOML.Buildpack.ID
```
```suggestion
					buildpackID := "A"
```

Comment (natalieparellano) [reply buildpack/build_test.go:236]: (this is what is actually used)

File: buildpack/build_test.go [hunk 12]
```diff
@@ -229,36 +199,113 @@ func testBuild(t *testing.T, when spec.G, it spec.S) {
...
+
+					h.Mkdir(t,
+						filepath.Join(layersDir, buildpackID))
+					h.Mkfile(t, `{"key": "some-bom-content"}`,
+						filepath.Join(layersDir, buildpackID, "launch.bom.cdx.json"),
+						filepath.Join(layersDir, buildpackID, "build.bom.cdx.json"),
+						filepath.Join(layersDir, buildpackID, fmt.Sprintf("%s.bom.cdx.json", layerName)))
+
+					h.Mkdir(t,
+						filepath.Join(layersDir, "A", layerName))
```
Comment (natalieparellano) [inline buildpack/build_test.go:247]:
```suggestion
						filepath.Join(layersDir, buildpackID, layerName))
```
```suggestion
						filepath.Join(layersDir, "A", layerName))
```

File: buildpack/build_test.go [hunk 13]
```diff
@@ -229,36 +199,113 @@ func testBuild(t *testing.T, when spec.G, it spec.S) {
...
-						Processes:   []launch.Process{},
-						Slices:      []layers.Slice{},
-					}); s != "" {
-						t.Fatalf("Unexpected:\n%s\n", s)
+					}, br)
+				})
+
+				it("does not process BOM files for old BP API versions", func() {
+					bpTOML.API = api.MustParse("0.2").String()
+					buildpackID := "A"
```
Comment (natalieparellano) [inline buildpack/build_test.go:288]:
```suggestion
					buildpackID := bpTOML.Buildpack.ID
```
```suggestion
					buildpackID := "A"
```

File: buildpack/build_test.go [hunk 14]
```diff
@@ -229,36 +199,113 @@ func testBuild(t *testing.T, when spec.G, it spec.S) {
...
+
+					h.Mkdir(t,
+						filepath.Join(layersDir, buildpackID))
+					h.Mkfile(t, `{"key": "some-bom-content"}`,
+						filepath.Join(layersDir, buildpackID, "launch.bom.cdx.json"),
+						filepath.Join(layersDir, buildpackID, "build.bom.cdx.json"),
+						filepath.Join(layersDir, buildpackID, fmt.Sprintf("%s.bom.cdx.json", layerName)))
+
+					h.Mkdir(t,
+						filepath.Join(layersDir, "A", layerName))
```
Comment (natalieparellano) [inline buildpack/build_test.go:299]:
```suggestion
						filepath.Join(layersDir, buildpackID, layerName))
```
```suggestion
						filepath.Join(layersDir, "A", layerName))
```

File: builder_test.go [hunk 15]
```diff
@@ -168,6 +169,86 @@ func testBuilder(t *testing.T, when spec.G, it spec.S) {
 				}
 			})
 
+			it("copies any created BOM files to the correct locations", func() {
+				bpA := testmock.NewMockBuildpack(mockCtrl)
+				bpB := testmock.NewMockBuildpack(mockCtrl)
+				buildpackStore.EXPECT().Lookup("A", "v1").Return(bpA, nil)
+				buildpackStore.EXPECT().Lookup("B", "v2").Return(bpB, nil)
+
+				bomFilePath1 := filepath.Join(layersDir, "launch.bom.cdx.json")
```
Comment (natalieparellano) [inline builder_test.go:178]: Should we also test that build and cache bomfiles are copied?

File: restorer.go [hunk 16]
```diff
@@ -93,16 +100,47 @@ func (r *Restorer) Restore(cache Cache) error {
...
+	if api.MustParse(r.Platform.API()).AtLeast("0.8") {
+		if r.LayersMetadata.BOM != nil && r.LayersMetadata.BOM.SHA != "" {
+			g.Go(func() error {
+				return r.restorePreviousLayer(r.LayersMetadata.BOM.SHA, previousImage)
+			})
+		}
+
+		if cacheMeta.BOM.SHA != "" {
+			g.Go(func() error {
+				return r.restoreLayer(cache, cacheMeta.BOM.SHA)
```
Comment (natalieparellano) [inline restorer.go:113]:
```suggestion
				return r.restoreCacheLayer(cache, cacheMeta.BOM.SHA)
```
```suggestion
				return r.restoreLayer(cache, cacheMeta.BOM.SHA)
```

File: restorer.go [hunk 17]
```diff
@@ -117,3 +155,72 @@ func (r *Restorer) restoreLayer(cache Cache, sha string) error {
 
 	return layers.Extract(rc, "")
 }
+
+func (r *Restorer) restoreSBOM() error {
+	if api.MustParse(r.Platform.API()).LessThan("0.8") {
```
Comment (natalieparellano) [inline restorer.go:160]: Should we move this gate into the calling function?

File: restorer_test.go [hunk 18]
```diff
@@ -309,20 +320,32 @@ func testRestorerBuilder(buildpackAPI, platformAPI string) func(t *testing.T, wh
...
+`, bomLayerSHA, cacheFalseLayerSHA, cacheLaunchLayerSHA, cacheOnlyLayerSHA, noGroupLayerSHA, escapedLayerSHA)
 
 					err = ioutil.WriteFile(
 						filepath.Join(cacheDir, "committed", "io.buildpacks.lifecycle.cache.metadata"),
 						[]byte(contents),
 						0600,
 					)
 					h.AssertNil(t, err)
+
+					h.Mkdir(t, filepath.Join(layersDir, "bom-buildpack.id"))
```
Comment (natalieparellano) [inline restorer_test.go:332]: Is this still needed?

File: restorer_test.go [hunk 19]
```diff
@@ -309,20 +320,32 @@ func testRestorerBuilder(buildpackAPI, platformAPI string) func(t *testing.T, wh
...
 				})
 
 				it.After(func() {
 					h.AssertNil(t, os.RemoveAll(tarTempDir))
 				})
 
+				it("restores any BOM files to the correct locations", func() {
+					h.SkipIf(t, api.MustParse(platformAPI).LessThan("0.8"), "PLATFORM API <0.8 does not restore sBOM")
```
Comment (natalieparellano) [inline restorer_test.go:340]:
```suggestion
					h.SkipIf(t, api.MustParse(platformAPI).LessThan("0.8"), "Platform API < 0.8 does not restore sBOM")
```
This will make my OCD happy...
```suggestion
					h.SkipIf(t, api.MustParse(platformAPI).LessThan("0.8"), "PLATFORM API <0.8 does not restore sBOM")
```

File: restorer_test.go [hunk 20]
```diff
@@ -634,8 +657,79 @@ func testRestorerBuilder(buildpackAPI, platformAPI string) func(t *testing.T, wh
...
+					h.AssertNil(t, os.RemoveAll(filepath.Join(layersDir, "sbom")))
+				})
+
+				it.After(func() {
+					h.AssertNil(t, fakeAppImage.Cleanup())
+					h.AssertNil(t, os.RemoveAll(artifactsDir))
+				})
+
+				it("restores the previous image SBOM layer", func() {
+					h.SkipIf(t, api.MustParse(platformAPI).LessThan("0.8"), "PLATFORM API <0.8 does not restore sBOM")
```
Comment (natalieparellano) [inline restorer_test.go:723]:
```suggestion
					h.SkipIf(t, api.MustParse(platformAPI).LessThan("0.8"), "Platform API < 0.8 does not restore sBOM")
```
```suggestion
					h.SkipIf(t, api.MustParse(platformAPI).LessThan("0.8"), "PLATFORM API <0.8 does not restore sBOM")
```

Comment (natalieparellano) [reply restorer_test.go:723]: Maybe we should put the `SkipIf` in the before block, that we don't do unnecessary setup?

File: restorer_test.go [hunk 21]
```diff
@@ -309,20 +320,32 @@ func testRestorerBuilder(buildpackAPI, platformAPI string) func(t *testing.T, wh
...
 				})
 
+				it("restores any BOM files to the correct locations", func() {
+					h.SkipIf(t, api.MustParse(platformAPI).LessThan("0.8"), "PLATFORM API <0.8 does not restore sBOM")
+
+					h.AssertNil(t, restorer.Restore(testCache, nil))
+
+					got := h.MustReadFile(t, filepath.Join(layersDir, "buildpack.id", "cache-true.bom.cdx.json"))
+					want := `{"key": "some-bom-content"}`
+					h.AssertEq(t, string(got), want)
```
Comment (natalieparellano) [inline restorer_test.go:346]:
```suggestion
					h.AssertEq(t, string(got), want)
					h.AssertPathDoesNotExist(t, filepath.Join(layersDir, "sbom"))
```

File: buildpack/build_test.go [hunk 22]
```diff
@@ -229,36 +199,113 @@ func testBuild(t *testing.T, when spec.G, it spec.S) {
...
+							Processes: []launch.Process{},
+							Slices:    []layers.Slice{},
+						}); s != "" {
+							t.Fatalf("Unexpected:\n%s\n", s)
+						}
+						assertLogEntry(t, logHandler, "BOM table isn't supported in this buildpack api version. The BOM should be written to <layer>.bom.<ext>, launch.bom.<ext>, or build.bom.<ext>.")
+					})
+				})
+
+				it("process any created BOM files", func() {
```
[the remark this answers is not in the document]
Comment (natalieparellano) [reply buildpack/build_test.go:235]: IDK why there are two suggestions - the first one is what I intended!

**Comment by natalieparellano** [comment]
The code looks good! I'm going to do some manual verification against the A/C in the issues. We should look into the CI failure as it seems to be consistent (though might be unrelated to the changes).

**Comment by natalieparellano** [comment]
It's impressive that the diff is so readable given the scope of the changes :)

**Comment by natalieparellano** [comment]
@aemengo if you change `reference = "some-ref"` in acceptance/testdata/restorer/container/layers/analyzed.toml to `reference = "fake-registry.io/some-namespace/some-repo"` it might fix the restorer tests (as it will not try to reach Docker Hub).

File: builder.go (hunk 8 above)
Comment (aemengo) [reply builder.go]: _Fixed in latest commit._

File: buildpack/build_test.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (aemengo) [reply buildpack/build_test.go]: _Fixed in latest commit._

File: buildpack/build_test.go (hunk 9 above)
Comment (aemengo) [reply buildpack/build_test.go]: _Fixed in latest commit._

File: buildpack/build_test.go (hunk 11 above)
Comment (aemengo) [reply buildpack/build_test.go]: _Fixed in latest commit._

File: buildpack/build_test.go (hunk 12 above)
Comment (aemengo) [reply buildpack/build_test.go]: _Fixed in latest commit._

File: buildpack/build_test.go (hunk 14 above)
Comment (aemengo) [reply buildpack/build_test.go]: _Fixed in latest commit._

File: buildpack/build_test.go (hunk 13 above)
Comment (aemengo) [reply buildpack/build_test.go]: _Fixed in latest commit._

File: builder_test.go (hunk 15 above)
Comment (aemengo) [reply builder_test.go:178]: _Fixed in latest commit._

File: restorer.go (hunk 16 above)
Comment (aemengo) [reply restorer.go]: _Fixed in latest commit._

File: restorer.go (hunk 17 above)
Comment (aemengo) [reply restorer.go]: _Fixed in latest commit._

File: restorer_test.go (hunk 18 above)
Comment (aemengo) [reply restorer_test.go]: _Fixed in latest commit._

File: restorer_test.go (hunk 19 above)
Comment (aemengo) [reply restorer_test.go]: _Fixed in latest commit._

File: restorer_test.go (hunk 21 above)
Comment (aemengo) [reply restorer_test.go:344]: _Fixed in latest commit._

File: buildpack/build.go [hunk 23]
```diff
@@ -289,6 +290,14 @@ func (b *Descriptor) readOutputFiles(bpLayersDir, bpPlanPath string, bpPlanIn Pl
 		}
 		br.MetRequires = names(bpPlanIn.filter(bpBuild.Unmet).Entries)
 
+		// set BOM files
+		if api.MustParse(b.API).AtLeast("0.7") {
```
Comment (jabrown85) [inline buildpack/build.go:294]: Is this version supposed to be `0.8`? I would have expected it to match here: https://github.com/buildpacks/lifecycle/pull/749/files#diff-5dce70063c4b1b07e611b3fcd32da0445a0b95970028c8992beb9c955e38320fR110

File: restorer.go [hunk 24]
```diff
@@ -88,22 +95,57 @@ func (r *Restorer) Restore(cache Cache) error {
...
 
 func (r *Restorer) restoresLayerMetadata() bool {
 	return api.MustParse(r.Platform.API()).AtLeast("0.7")
 }
 
-func (r *Restorer) restoreLayer(cache Cache, sha string) error {
+func (r *Restorer) restorePreviousLayer(sha string, image imgutil.Image) error {
+	// Sanity check to prevent panic.
+	if image == nil {
+		return errors.New("restoring layer: previous image not found")
```
Comment (jabrown85) [inline restorer.go:136]:
```suggestion
		return errors.New(fmt.Sprintff("restoring layer: previous image not found for %q, sha"))
```

File: cmd/lifecycle/restorer.go [hunk 25]
```diff
@@ -126,12 +160,32 @@ func (r restoreArgs) restore(layerMetadata platform.LayersMetadata, group buildp
 		LayersMetadata:        layerMetadata,
 	}
 
-	if err := restorer.Restore(cacheStore); err != nil {
+	if err := restorer.Restore(cacheStore, previousImage); err != nil {
```
Comment (jabrown85) [inline cmd/lifecycle/restorer.go:163]: What was the reason we didn't sort of follow the `LayerMetadataRestorer` pattern with `SBOMFileRestorer`?

**Comment by aemengo** [comment]
> @aemengo if you change `reference = "some-ref"` in acceptance/testdata/restorer/container/layers/analyzed.toml to `reference = "fake-registry.io/some-namespace/some-repo"` it might fix the restorer tests (as it will not try to reach Docker Hub).

@natalieparellano I don't think that this did anything.

File: restorer.go [hunk 26]
```diff
@@ -117,3 +159,68 @@ func (r *Restorer) restoreLayer(cache Cache, sha string) error {
...
+	var bomRegex *regexp.Regexp
+
+	if runtime.GOOS == "windows" {
+		bomRegex = regexp.MustCompile(fmt.Sprintf(`%s\\(.+)\\(.+)\\(bom.+json)`, bomType))
+	} else {
+		bomRegex = regexp.MustCompile(fmt.Sprintf(`%s/(.+)/(.+)/(bom.+json)`, bomType))
+	}
+
+	return func(path string, info fs.FileInfo, err error) error {
+		if info != nil && info.IsDir() {
```
Comment (jabrown85) [inline restorer.go:194]: Should we verify this is a file and not a symlink? Just thinking about potential edge cases

**Comment by natalieparellano** [comment]
@aemengo we probably need to spin up a test registry and put a "real" previous image fake there, like we do for analyzer acceptance.

File: buildpack/build.go (hunk 23 above)
Comment (natalieparellano) [reply buildpack/build.go:294]: This is the buildpack API - the sBOM work requires both a platform and buildpack api bump.

File: restorer.go (hunk 24 above)
Comment (aemengo) [reply restorer.go]: _Fixed in latest commit._

File: cmd/lifecycle/restorer.go (hunk 25 above)
Comment (aemengo) [reply cmd/lifecycle/restorer.go:163]: **My humble opinion:**
I didn't think the complexity introduced in `restorer.go` was worth the tradeoff of hiding behind another abstraction. As a newcomer to the code, it took me quite a bit of time to understand what the `LayerMetadataRestorer` was doing exactly, and how it related to the rest of the code. I was trying to avoid the same with this BOM restoration code.

But the biggest reason was the performance implications. The BOM restoration code makes 2 asynchronous calls to restore layers, which happens at the same time that buildpack layers are being restored.

```go
				g.Go(func() error {
					return r.restoreCacheLayer(cache, cachedLayer.SHA)
				})
```

If we moved these calls to separate abstractions, it would be hard/clumsy to keep sure that they're having in parallel, for future developers

File: restorer.go (hunk 26 above)
Comment (aemengo) [reply restorer.go]: _Fixed in latest commit_

File: restorer_test.go (hunk 20 above)
Comment (aemengo) [reply restorer_test.go]: Fixed in latest commit

File: restorer.go (hunk 26 above)
Comment (jabrown85) [reply restorer.go]: nit: it is a bit harder to review changes when force pushing since the PR diff expands the entire file each time. Is there a better way for me to see changes as they come in?

**Comment by natalieparellano** [comment]
Some manual verification done using the creator task/taskrun from Tekton:

Using a builder built from the following branch of the samples repo: https://github.com/buildpacks/samples/compare/sbom?expand=1

The exported image contains the bom layer sha in `io.buildpacks.lifecycle.metadata`:
```
$  docker inspect 192.168.1.43:5000/test-sbom-03 | jq .[0].Config.Labels
{
  "io.buildpacks.build.metadata": "{\"bom\":null,\"buildpacks\":[{\"homepage\":\"https://github.com/buildpacks/samples/tree/main/buildpacks/java-maven\",\"id\":\"samples/java-maven\",\"version\":\"0.0.1\"}],\"launcher\":{\"version\":\"0.12.0-rc.1-15+e1f808b\",\"source\":{\"git\":{\"repository\":\"github.com/buildpacks/lifecycle\",\"commit\":\"e1f808b\"}}},\"processes\":[{\"type\":\"web\",\"command\":\"java -jar target/sample-0.0.1-SNAPSHOT.jar\",\"args\":null,\"direct\":false,\"buildpackID\":\"samples/java-maven\"}]}",
  "io.buildpacks.lifecycle.metadata": "{\"app\":[{\"sha\":\"sha256:bfaec016a54782bb3ade6dbe8a8377fbfbd8dbca738ca13efda26c29e7394fb6\"}],\"bom\":{\"sha\":\"sha256:243ec16aee04f8b1e68469d7937b71fc5f58619dbd67e75a14f2bf9aefa3565c\"},\"buildpacks\":[{\"key\":\"samples/java-maven\",\"version\":\"0.0.1\",\"layers\":{\"jdk\":{\"sha\":\"sha256:224174ad98acdfd53380f92aedd6dd33885125e4e06fb45743a79b0d9651ea39\",\"data\":{\"url\":\"https://cdn.azul.com/zulu/bin/zulu8.40.0.25-ca-jdk8.0.222-linux_x64.tar.gz\",\"version\":\"1.8.0_222\"},\"build\":true,\"launch\":true,\"cache\":true}}}],\"config\":{\"sha\":\"sha256:db36a65d3fccec59a34312ef7fd572ddd4bcddae5d073f0d91f06435d0d36aec\"},\"launcher\":{\"sha\":\"sha256:54a0873ca22cd93cf6606731910e11696c6c6fad21990168cd4c1d1ddd95f73d\"},\"process-types\":{\"sha\":\"sha256:83d85471d9f8a3834b4e27cf701e3f0aef220cc816d9c173c7d32cd73909a590\"},\"runImage\":{\"topLayer\":\"sha256:5933667ff857157c54b874d8ace62126540242f05b5331b94053b62f7d86c55b\",\"reference\":\"index.docker.io/cnbs/sample-stack-run@sha256:38f19fee4ffd8caafb5550da2d26687feba159528d041101ad99cc7390a0f578\"},\"stack\":{\"runImage\":{\"image\":\"cnbs/sample-stack-run:bionic\"}}}",
  "io.buildpacks.project.metadata": "{}",
  "io.buildpacks.stack.id": "io.buildpacks.samples.stacks.bionic"
}
```

The bom layer sha contains the expected data:
Screen Shot 2021-11-02 at 3 05 41 PM

```
$  docker run --rm -it --entrypoint /bin/bash 192.168.1.43:5000/test-sbom-03  
cnb@9261184954b9:/workspace/source/apps/java-maven$ cat /layers/sbom/launch/samples_java-maven/bom.cdx.json 
launch_var: "val"
cnb@9261184954b9:/workspace/source/apps/java-maven$ cat /layers/sbom/launch/samples_java-maven/jdk/bom.cdx.json 
jdk_var: "val"

```

On a rebuild, the restored bom files are cleaned from /layers/sbom before buildpack execution:
```
===> RESTORING
Restoring metadata for "samples/java-maven:jdk" from app image
Restoring metadata for "samples/java-maven:maven_m2" from cache
Restoring data for "samples/java-maven:jdk" from cache
Restoring data for "samples/java-maven:maven_m2" from cache
===> BUILDING
---> Java buildpack
Checking layers
total 12
drwxrwxrwx 3 cnb  cnb  4096 Nov  2 19:10 .
drwxr-xr-x 1 root root 4096 Nov  2 19:10 ..
drwxr-xr-x 4 cnb  cnb  4096 Nov  2 19:10 samples_java-maven
ls: cannot access '/layers/sbom': No such file or directory
ls: cannot access '/layers/sbom/build': No such file or directory
Checking bp layers
ls: cannot access '/layers/sbom/cache': No such file or directory
ls: cannot access '/layers/sbom/launch': No such file or directory
```

❗ On a rebuild, the restored bom files are NOT copied to the bp layers dir:
```
drwxr-xr-x  4 cnb cnb 4096 Nov  2 19:10 .
drwxrwxrwx  3 cnb cnb 4096 Nov  2 19:10 ..
drwxr-xr-x 11 cnb cnb 4096 Nov  2 19:10 jdk
-rw-r--r--  1 cnb cnb  120 Nov  2 19:10 jdk.toml
drwxr-xr-x  4 cnb cnb 4096 Nov  2 19:10 maven_m2
-rw-r--r--  1 cnb cnb    0 Nov  2 19:10 maven_m2.toml
```

On a rebuild, the bom launch and cache layers are re-used if they didn't change:
```
===> EXPORTING
Reusing layers from image '192.168.1.43:5000/test-sbom-03@sha256:888b9136047d74c385613ac7e949f84a37e3d56c7c74bef067a7cf6afbff5476'
Reusing layer 'samples/java-maven:jdk'
Reusing layer 'launch.bom'
Adding 1/1 app layer(s)
Reusing layer 'launcher'
Reusing layer 'config'
Reusing layer 'process-types'
Adding label 'io.buildpacks.lifecycle.metadata'
Adding label 'io.buildpacks.build.metadata'
Adding label 'io.buildpacks.project.metadata'
Setting default process type 'web'
Saving 192.168.1.43:5000/test-sbom-03...
*** Images (sha256:8504eef668eb13afab9e04fc6bc6987a07bede51f401b8894114b327d91d0027):
      192.168.1.43:5000/test-sbom-03
Reusing cache layer 'samples/java-maven:jdk'
Adding cache layer 'samples/java-maven:maven_m2'
Reusing cache layer 'cache.bom'
```

File: restorer.go (hunk 26 above)
Comment (aemengo) [reply restorer.go]: If you have ideas, I'm happy to improve my PR etiquette. Conversely if I had made a new commit for every review comment, I'd imagine it would be much harder to digest the history, and much harder for me to keep up-to-date.

**Comment by aemengo** [comment]
@natalieparellano Unfortunately, I've been unable to reproduce. Let's please look at this together

```
===> DETECTING
elm-buildpack 0.0.1
===> RESTORING
Restoring metadata for "elm-buildpack:launch-only" from app image
Restoring metadata for "elm-buildpack:launch-true-cache-true" from app image
Restoring data for "elm-buildpack:launch-true-cache-true" from cache
===> BUILDING
---> Showing Layers Directory
/layers/elm-buildpack
|-- launch-only.bom.cdx.json             # <---
|-- launch-only.toml
|-- launch-true-cache-true
|   `-- elm
|-- launch-true-cache-true.bom.cdx.json  # <---
`-- launch-true-cache-true.toml
```

File: restorer.go [hunk 27]
```diff
@@ -117,3 +158,68 @@ func (r *Restorer) restoreLayer(cache Cache, sha string) error {
...
+			return nil
+		}
+
+		return Copy(path, dest)
+	}
+}
+
+func (r *Restorer) buildpackDetected(id string) bool {
+	for _, bp := range r.Buildpacks {
+		if bp.ID == id {
```
[the remark this answers is not in the document]
Comment (aemengo) [reply restorer.go]: This is a good catch! _Fixed in latest commit._

File: analyzer.go [hunk 28]
```diff
@@ -56,18 +65,24 @@ func (a *Analyzer) Analyze() (platform.AnalyzedMetadata, error) {
...
+	}
 
+	if a.restoresLayerMetadata() {
 		useShaFiles := true
 		if err := a.LayerMetadataRestorer.Restore(a.Buildpacks, appMeta, cacheMeta, NewLayerSHAStore(useShaFiles)); err != nil {
 			return platform.AnalyzedMetadata{}, err
 		}
 	}
 
+	if cacheMeta.BOM.SHA != "" {
```
Comment (natalieparellano) [inline analyzer.go:80]: Should we also check the platform API here?

**Review by natalieparellano — COMMENTED** [review]
Looks good to me! I think we should answer the question [here](https://github.com/buildpacks/lifecycle/pull/749#discussion_r733989124). But otherwise I am 👍🏼  on this!

File: analyzer.go (hunk 28 above)
Comment (aemengo) [reply analyzer.go]: Yes we should. This was an oversight by me. _Fixed._

File: buildpack/bomfile.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (aemengo) [reply buildpack/bomfile.go:58]: @samj1912 I apologize. It seems that you were one step ahead of me, as usual. Could you please take a look at my latest commit and tell me what you think?

**Review by sclevine — APPROVED** [review]

**Review by natalieparellano — APPROVED** [review]
Amazing work @aemengo !!!

File: buildpack/bomfile.go [hunk 29]
```diff
@@ -13,6 +15,13 @@ const (
 	LayerTypeLaunch
 )
 
+const (
+	mediaTypeCycloneDX   = "application/vnd.cyclonedx+json"
+	mediaTypeSPDX        = "text/spdx+json"
```
Comment (samj1912) [inline buildpack/bomfile.go:20]: This ended up being registered as `application/spdx+json` per https://github.com/spdx/spdx-spec/issues/577#issuecomment-960295523

Comment (aemengo) [reply buildpack/bomfile.go]: Thanks. _Fixed in latest commit._

**Review by jabrown85 — APPROVED** [review]

File: buildpack/bomfile.go [hunk 30]
```diff
@@ -0,0 +1,160 @@
...
+
+func (b *BOMFile) mediaType() string {
+	name := filepath.Base(b.Path)
+
+	switch {
+	case strings.HasSuffix(name, "bom.cdx.json"):
+		return mediaTypeCycloneDX
+	case strings.HasSuffix(name, "bom.spdx.json"):
+		return mediaTypeSPDX
+	case strings.HasSuffix(name, "bom.syft.json"):
```
Comment (sclevine) [inline buildpack/bomfile.go:61]:
```suggestion
	case strings.HasSuffix(name, ".bom.syft.json"):
```

**Comment by samj1912** [comment]
Just noting this from the Core team WG meeting today - we decided to be consistent about the naming and use `sbom` instead of `bom` everywhere.

**Comment by natalieparellano** [comment]
> Just noting this from the Core team WG meeting today - we decided to be consistent about the naming and use sbom instead of bom everywhere.

I am going to merge this PR with a fast-follow PR that makes the bom -> sbom change.

File: buildpack/bomfile.go (hunk 30 above)
Comment (natalieparellano) [reply buildpack/bomfile.go:61]: Addressed in https://github.com/buildpacks/lifecycle/pull/759
